include/message_filters/sync_policies/exact_time.hpp

```diff
@@ -154,40 +154,42 @@ struct ExactTime : public PolicyBase<Ms...>
       clearOldTuples();
     }
 
-    if (queue_size_ > 0) {
-      while (tuples_.size() > queue_size_) {
-        Tuple & t2 = tuples_.begin()->second;
-        std::apply([this](auto &&... args) {this->drop_signal_.call(args ...);}, t2);
-        tuples_.erase(tuples_.begin());
-      }
+    if (queue_size_ > 0 && tuples_.size() > queue_size_) {
+      const auto n_exceeding_tuples = tuples_.size() - queue_size_;
+      const auto oldest_to_keep = std::next(tuples_.begin(), n_exceeding_tuples);
+      drop_range(tuples_.begin(), oldest_to_keep);
     }
   }
 
+
   // assumes mutex_ is already locked
   void clearOldTuples()
   {
-    typename M_TimeToTuple::iterator it = tuples_.begin();
-    typename M_TimeToTuple::iterator end = tuples_.end();
-    for (; it != end; ) {
-      if (it->first <= last_signal_time_) {
-        typename M_TimeToTuple::iterator old = it;
-        ++it;
-
-        Tuple & t = old->second;
-        std::apply([this](auto &&... args) {this->drop_signal_.call(args ...);}, t);
-        tuples_.erase(old);
-      } else {
-        // the map is sorted by time, so we can ignore anything after this if this one's time is ok
-        break;
-      }
-    }
+    const auto arrived_before_last_signal = [this](const auto & e)
+      {return e.first <= last_signal_time_;};
+    const auto first_arrived_after =
+      std::partition_point(tuples_.begin(), tuples_.end(), arrived_before_last_signal);
```

Comment:
include `<algorithm>`

Reply:
Not needed anymore: it turns out, that `std::map::upper_bound` is a way better solution in this case.